File: src/modules/algorithm/seed/algorithms/queue/languages/typescript/linkedlist-solution.ts
class Node {
  data: any;
  next: Node | null;

  constructor(data: any) {
    this.data = data;
    this.next = null;
  }
}

export class Queue {
  private front: Node | null;
  private rear: Node | null;
  private count: number;
  private capacity: number;

  constructor(capacity: number = Infinity) {
    this.front = null;
    this.rear = null;
    this.count = 0;
    this.capacity = capacity;
  }

  enqueue(item: any): void {
    if (this.count >= this.capacity) {
      throw new Error('Queue Overflow');
    }

    const newNode = new Node(item);

    if (this.isEmpty()) {
      this.front = newNode;
      this.rear = newNode;
    } else {
      if (this.rear) {
        this.rear.next = newNode;
      }
      this.rear = newNode;
    }

    this.count++;
  }

  dequeue(): any {
    if (this.isEmpty()) {
      return undefined;
    }

    const item = this.front!.data;
    this.front = this.front!.next;
    this.count--;

    if (this.isEmpty()) {
      this.rear = null;
    }

    return item;
  }

  peek(): any {
    if (this.isEmpty()) {
      return undefined;
    }
    return this.front!.data;
  }

  isEmpty(): boolean {
    return this.count === 0 && this.front === null;
  }

  isFull(): boolean {
    return this.count === this.capacity;
  }

  size(): number {
    return this.count;
  }
}

File: src/modules/algorithm/seed/algorithms/queue/languages/python/linkedlist-solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self, capacity: int = float('inf')):
        """
        Initializes an empty queue with a given capacity.
        """
        self.front = None
        self.rear = None
        self.count = 0
        self.capacity = capacity

    def enqueue(self, item: any) -> None:
        """
        Adds an element to the rear of the queue.
        """
        if self.count >= self.capacity:
            raise OverflowError("Queue is full")
        new_node = Node(item)
        if self.is_empty():
            self.front = new_node
            self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node
        self.count += 1

    def dequeue(self) -> any:
        """
        Removes an element from the front of the queue.
        """
        if self.is_empty():
            return None
        item = self.front.data
        self.front = self.front.next
        self.count -= 1
        if self.is_empty():
            self.rear = None
        return item

    def peek(self) -> any:
        """
        Returns the element at the front of the queue without removing it.
        """
        if self.is_empty():
            return None
        return self.front.data

    def is_empty(self) -> bool:
        """
        Checks if the queue is empty.
        """
        return self.count == 0

    def is_full(self) -> bool:
        """
        Checks if the queue is full.
        """
        return self.count == self.capacity

    def size(self) -> int:
        """
        Returns the number of elements in the queue.
        """
        return self.count 